scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -250,25 +250,52 @@ def highly_variable_genes(
             'pass `inplace=False` if you want to return a `np.recarray`.')
 
     if batch_key is None:
-        df = _highly_variable_genes_single_batch(adata,
-                                                 min_disp=min_disp, max_disp=max_disp,
-                                                 min_mean=min_mean, max_mean=max_mean,
-                                                 n_top_genes=n_top_genes,
-                                                 n_bins=n_bins,
-                                                 flavor=flavor)
+        df = _highly_variable_genes_single_batch(
+            adata,
+            min_disp=min_disp, max_disp=max_disp,
+            min_mean=min_mean, max_mean=max_mean,
+            n_top_genes=n_top_genes,
+            n_bins=n_bins,
+            flavor=flavor,
+        )
     else:
         sanitize_anndata(adata)
         batches = adata.obs[batch_key].cat.categories
         df = []
         for batch in batches:
             adata_subset = adata[adata.obs[batch_key] == batch]
-            hvg = _highly_variable_genes_single_batch(adata_subset,
-                                                      min_disp=min_disp, max_disp=max_disp,
-                                                      min_mean=min_mean, max_mean=max_mean,
-                                                      n_top_genes=n_top_genes,
-                                                      n_bins=n_bins,
-                                                      flavor=flavor)
-            hvg['gene'] = adata.var_names.values
+
+            # Filter to genes that are in the dataset
+            genes_before = adata_subset.var_names
+            filter_genes(adata_subset, min_cells=1)
+            genes_after = adata_subset.var_names
+            genes_missing = genes_before.difference(genes_after)
+
+            hvg = _highly_variable_genes_single_batch(
+                adata_subset,
```

Comment:
This function should probably take an array/ sparse matrix directly, since it doesn't actually use anything else from the `AnnData`.

Reply:
Hmm... I didn't make that function. That was done by @gokceneraslan when he added `batch_key`. I would therefore be inclined to leave it like that in this PR.